library/src/interactor_impl.cxx

```diff
@@ -71,11 +75,85 @@ class interactor_impl::internals
 #endif
   }
 
+  //----------------------------------------------------------------------------
+  // Method defined to normalize the Z axis so all models are treated temporarily
+  // as Z-up axis models. 
+  static vtkNew<vtkTransform> zUpTransform(vtkRenderer* renderer, vtkTransform* transform)
+  {
+    const double* up = renderer->GetEnvironmentUp();
+    const double* right = renderer->GetEnvironmentRight();
+    double fwd[3];
+    vtkMath::Cross(right, up, fwd);
+
+    const double m[16] = {
+      right[0], right[1], right[2], 0, //
+      fwd[0],   fwd[1],   fwd[2],   0, //
+      up[0],    up[1],    up[2],    0, //
+      0,        0,        0,        1, //
+    };
+    vtkNew<vtkTransform> toZup;
+    toZup->SetMatrix(m);
+    return toZup;
+  }
```

Comment:
what @mwestphal meant was to use an out parameter instead of a return
```suggestion
  static void zUpTransform(vtkRenderer* renderer, vtkTransform* transform)
  {
    const double* up = renderer->GetEnvironmentUp();
    const double* right = renderer->GetEnvironmentRight();
    double fwd[3];
    vtkMath::Cross(right, up, fwd);

    const double m[16] = {
      right[0], right[1], right[2], 0, //
      fwd[0],   fwd[1],   fwd[2],   0, //
      up[0],    up[1],    up[2],    0, //
      0,        0,        0,        1, //
    };
    transform->SetMatrix(m);
  }
```

---

But if we're going that route we may want to go all out and do
```suggestion
  static void zUpTransforms(vtkRenderer* renderer, vtkTransform* to, vtkTransform* from)
  {
    const double* up = renderer->GetEnvironmentUp();
    const double* right = renderer->GetEnvironmentRight();
    double fwd[3];
    vtkMath::Cross(right, up, fwd);

    const double m[16] = {
      right[0], right[1], right[2], 0, //
      fwd[0],   fwd[1],   fwd[2],   0, //
      up[0],    up[1],    up[2],    0, //
      0,        0,        0,        1, //
    };
    to->SetMatrix(m);
    from->SetMatrix(m);
    from->Inverse();
  }
```
calling as:
```cpp
        vtkRenderer* renderer = self->VTKInteractor->GetRenderWindow()->GetRenderers()->GetFirstRenderer();
        vtkNew<vtkTransform> toZup, fromZup;
        zUpTransform(renderer, toZup, fromZup);
```

---

but wait! there's more! if the function is in the class, why not make it a member instead of static and grab the renderer from inside?
```suggestion
  void zUpTransforms(vtkTransform* to, vtkTransform* from)
  {
    vtkRenderer* renderer = this->VTKInteractor->GetRenderWindow()->GetRenderers()->GetFirstRenderer();
    const double* up = renderer->GetEnvironmentUp();
    const double* right = renderer->GetEnvironmentRight();
    double fwd[3];
    vtkMath::Cross(right, up, fwd);

    const double m[16] = {
      right[0], right[1], right[2], 0, //
      fwd[0],   fwd[1],   fwd[2],   0, //
      up[0],    up[1],    up[2],    0, //
      0,        0,        0,        1, //
    };
    to->SetMatrix(m);
    from->SetMatrix(m);
    from->Inverse();
  }
```
```cpp
        vtkNew<vtkTransform> toZup, fromZup;
        self->zUpTransforms(toZup, fromZup);
```

Reply:
make sense to me :)